icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1386,53 +1442,59 @@ private boolean replaceLanguage(boolean checkLanguage,
+                String replacedLanguage = null;
+
+                if (replacement.indexOf('_') < 0) {
+                    replacedLanguage = replacement.equals(UNDEFINED_LANGUAGE) ?
+                        language : replacement;
+                } else {
+                    String[] replacementFields = replacement.split("_");
+                    replacedLanguage = replacementFields[0];
+                    int index = 1;
+
+                    if (replacedLanguage.equals(UNDEFINED_LANGUAGE)) {
+                        replacedLanguage = language;
+                    }
+                    int consumed = replacementFields[0].length() + 1;
+                    while (replacementFields.length > index) {
+                        String field = replacementFields[index];
+                        int len = field.length();
+                        if (1 == len) {
+                            replacedExtensions = replacement.substring(consumed);
+                            break;
+                        } else if (len >= 2 && len <= 3) {
+                            assert replacedRegion == null;
+                            replacedRegion = field;
+                        } else if (len >= 5 && len <= 8) {
                             assert replacedVariant == null;
                             replacedVariant = field;
-                        } else {
-                            assert replacedScript == null;
-                            replacedScript = field;
+                        } else if (len == 4) {
+                            if (field.charAt(0) >= '0' && field.charAt(0) <= '9') {
+                                assert replacedVariant == null;
+                                replacedVariant = field;
+                            } else {
+                                assert replacedScript == null;
+                                replacedScript = field;
+                            }
                         }
+                        index++;
+                        consumed += len + 1;
                     }
-                    index++;
-                    consumed += len + 1;
                 }
 
                 replacedScript = deleteOrReplace(script, null, replacedScript);
                 replacedRegion = deleteOrReplace(region, searchRegion, replacedRegion);
                 replacedVariant = deleteOrReplace(searchVariant, searchVariant, replacedVariant);
 
-                if (same(this.language, replacedLanguage) &&
-                    same(this.script, replacedScript) &&
-                    same(this.region, replacedRegion) &&
-                    same(searchVariant, replacedRegion) &&
-                    replacedExtensions == null) {
+                if (    this.language.equals(replacedLanguage) &&
+                        this.script.equals(replacedScript) &&
+                        this.region.equals(replacedRegion) &&
+                        searchVariant.equals(replacedVariant) &&
```

Comment:
can be null, right? if so, then please use `Objects.equals(searchVariant, replacedVariant)` to avoid a crash and handle nulls